### PR TITLE
Add response metadata to response header for partitionAssignment

### PR DESCRIPTION
**Pull request by xyuanlu**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:

#1746

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

User may call this API asynchronously. We should return some metadata of the response to user as well.  

### Tests

- [X] The following tests are written for this issue:

testComputePartitionAssignment

- The following is the result of the "mvn test" command on the appropriate module:
```
[INFO] Tests run: 179, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 325.848 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 179, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.6:report (generate-code-coverage-report) @ helix-rest ---
[INFO] Loading execution data file /Users/xialu/Documents/WorkSpace/helix/helix-rest/target/jacoco.exec
[INFO] Analyzed bundle 'Apache Helix :: Restful Interface' with 79 classes
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  05:34 min
[INFO] Finished at: 2021-07-21T14:51:27-07:00
[INFO] ------------------------------------------------------------------------
```

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

**Review by junkaixue — COMMENTED** [review]
I have a general question, what is the reason to have the header value in the response?

**Comment by xyuanlu** [comment]
> I have a general question, what is the reason to have the header value in the response?

Thanks for the question.
When user call this API in async mode, we need to add some metadata info in the reply indicating the return format/filter value etc. 
Usually metadata stores in header instead of payload.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/AbstractResource.java [hunk 1]
```diff
@@ -160,6 +164,16 @@ protected Response JSONRepresentation(Object entity) {
     }
   }
 
+  protected Response JSONRepresentation(Object entity, String headerName, Object headerValue) {
```
Comment (huizhilu) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/AbstractResource.java:167]: I suggest it should reuse the existing method `protected Response JSONRepresentation(Object entity)` to combine them, reducing duplicate code.

**Comment by junkaixue** [comment]
Sync up with @xyuanlu . Following actions needed:
1. Add comments for the suggestion of leveraging header to store metadata.
2. Think about what we can do that let existing customers know the header has info can be retrieved as well. I think most of our customers only relying on the payload returned by REST service.

**Comment by junkaixue** [comment]
@xyuanlu any update on this?

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/AbstractResource.java (hunk 1 above)
Comment (xyuanlu) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/AbstractResource.java:168]: TFTR. Updated.

**Comment by xyuanlu** [comment]
> @xyuanlu any update on this?

Thanks for the review.
I have updated the PR. I plan to update the REST API document after this PR is merged. And maybe propose to evaluate all current APIs to see if we pass any metadata in response body (that may need to be reconstructed).

Thanks!

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/AbstractResource.java [hunk 2]
```diff
@@ -130,6 +130,13 @@ protected Response OK() {
     return Response.ok().build();
   }
 
+  protected Response OKWithHeader(Object entity, String headerName, Object headerValue) {
+    if (headerName == null || headerName.length()==0)
+      return OK(entity);
+    else
```
Comment (junkaixue) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/AbstractResource.java:136]: NIT: remove the else and add {} to if statement. Let's respect our coding convention.

**Review by junkaixue — APPROVED** [review]
Overall, LGTM

**Comment by xyuanlu** [comment]
This change is ready to be merged. 
Approved  by @junkaixue 

Commit message:
"Add response metadata to response header for partitionAssignment"

Thanks!!

**Pull request merged**